### DISCUSSION TITLE
📚  Organizing stylelint recommended rules

### DISCUSSION THREAD

**Discussion by togami2864** [biomejs/biome#2422] [General]
I was organizing and reviewing whether it's necessary to implement [stylelint-config-recommended](https://github.com/stylelint/stylelint-config-recommended/blob/main/index.js). I have marked some rules with a `?` to indicate uncertainties or potential issues.

Should we consider using a parser or formatter instead of a linter for applying these rules?

| rule | Need? | Difficulties | Note |
| ---- | ------ | ------------ | ---- |
| [no-invalid-double-slash-comments](https://github.com/stylelint/stylelint/blob/14.8.2/lib/rules/no-invalid-double-slash-comments/README.md) | No | - | [parser covers this](https://biomejs.dev/playground/?files.main.css=YQAgAHsACgAgACAALwAvAGMAbwBsAG8AcgA6ACAAcABpAG4AawA7AAoAfQA%3D) |
| [annotation-no-unknown](https://github.com/stylelint/stylelint/blob/14.8.2/lib/rules/annotation-no-unknown/README.md)| ? | - | [parser reports error but a little imperfect](https://biomejs.dev/playground/?files.main.css=YQAgAHsACgAgACAAYwBvAGwAbwByADoAIABnAHIAZQBlAG4AIAAhAGkAbQBtAG0AbQBtAHAAbwByAHQAYQBuAHQAOwAKAH0A) |
| [color-no-invalid-hex](https://github.com/stylelint/stylelint/blob/14.8.2/lib/rules/color-no-invalid-hex/README.md) | ? | - | The parser covers some cases, but it's not perfect. |
| [function-calc-no-unspaced-operator](https://github.com/stylelint/stylelint/blob/14.8.2/lib/rules/function-calc-no-unspaced-operator/README.md) | ? | - | style rule? It would conflict with formatter? |
| [no-empty-source](https://github.com/stylelint/stylelint/blob/14.8.2/lib/rules/no-empty-source/README.md) | ? | - | syntax error occured https://stylelint.io/user-guide/rules/no-empty-source/ |
| [string-no-newline](https://github.com/stylelint/stylelint/blob/14.8.2/lib/rules/string-no-newline/README.md) | ? | - | [style rule?](https://biomejs.dev/playground/?files.main.css=YQAgAHsACgAgACAAYwBvAG4AdABlAG4AdAA6ACAAIgBmAGkAcgBzAHQACgAgACAAIAAgAHMAZQBjAG8AbgBkACIAOwAKAH0A) |
| [media-query-no-invalid](https://github.com/stylelint/stylelint/blob/14.8.2/lib/rules/media-query-no-invalid/README.md) | ? | - | [parser will panic](https://biomejs.dev/playground/?files.main.css=QABtAGUAZABpAGEAIAAoAHcAaQBkAHQAaAAgAD0APQAgADEAMAAwAHAAeAApACAAewB9AA%3D%3D) |

## Full List
(Difficulties are based on my perspective) 

| rule | Need? | Difficulties | Note |
| ---- | ------ | ------------ | ---- |
| [block-no-empty](https://github.com/stylelint/stylelint/blob/14.8.2/lib/rules/block-no-empty/README.md) | Yes | Easy | - |
|[comment-no-empty](https://github.com/stylelint/stylelint/blob/14.8.2/lib/rules/comment-no-empty/README.md) | Yes | Easy | - |
| [font-family-no-duplicate-names](https://github.com/stylelint/stylelint/blob/14.8.2/lib/rules/font-family-no-duplicate-names/README.md) | Yes | Easy | [Done](https://github.com/biomejs/biome/pull/2308) |
| [font-family-no-missing-generic-family-keyword](https://github.com/stylelint/stylelint/blob/14.8.2/lib/rules/font-family-no-missing-generic-family-keyword/README.md)| Yes | Easy | - |
| [function-no-unknown](https://github.com/stylelint/stylelint/blob/14.8.2/lib/rules/function-no-unknown/README.md) | Yes | Easy | - |
| [keyframe-block-no-duplicate-selectors](https://github.com/stylelint/stylelint/blob/14.8.2/lib/rules/keyframe-block-no-duplicate-select/README.md) | Yes | Easy | - |
| [keyframe-declaration-no-important](https://github.com/stylelint/stylelint/blob/14.8.2/lib/rules/keyframe-declaration-no-important/README.md) | Yes | Easy | - |
| [media-feature-name-no-unknown](https://github.com/stylelint/stylelint/blob/14.8.2/lib/rules/media-feature-name-no-unknown/README.md) | Yes | Easy | - |
| [no-duplicate-at-import-rules](https://github.com/stylelint/stylelint/blob/14.8.2/lib/rules/no-duplicate-at-import-rules/README.md) | Yes | Easy | - |
| [no-duplicate-selectors](https://github.com/stylelint/stylelint/blob/14.8.2/lib/rules/no-duplicate-selectors/README.md) | Yes | Easy | - |
| [no-irregular-whitespace](https://github.com/stylelint/stylelint/blob/14.8.2/lib/rules/no-irregular-whitespace/README.md) | Yes | Easy | - |
| [selector-pseudo-class-no-unknown](https://github.com/stylelint/stylelint/blob/14.8.2/lib/rules/selector-pseudo-class-no-unknown/README.md) | Yes | Easy | - |
| [selector-pseudo-element-no-unknown](https://github.com/stylelint/stylelint/blob/14.8.2/lib/rules/selector-pseudo-element-no-unknown/README.md) | Yes | Easy | - |
| [selector-type-no-unknown](https://github.com/stylelint/stylelint/blob/14.8.2/lib/rules/selector-type-no-unknown/README.md) | Yes | Easy | - |
| [unit-no-unknown](https://github.com/stylelint/stylelint/blob/14.8.2/lib/rules/unit-no-unknown/README.md) | Yes | Easy | - |
| [custom-property-no-missing-var-function](https://github.com/stylelint/stylelint/blob/14.8.2/lib/rules/custom-property-no-missing-var-function/README.md) | Yes | Mid | - |
| [declaration-block-no-duplicate-custom-properties](https://github.com/stylelint/stylelint/blob/14.8.2/lib/rules/declaration-block-no-duplicate-custom-properties/README.md) | Yes | Mid | - |
| [declaration-block-no-shorthand-property-overrides](https://github.com/stylelint/stylelint/blob/14.8.2/lib/rules/declaration-block-no-shorthand-property-overrides/README.md) | Yes | Mid | - |
| [function-linear-gradient-no-nonstandard-direction](https://github.com/stylelint/stylelint/blob/14.8.2/lib/rules/function-linear-gradient-no-nonstandard-direction/README.md) | Yes | Mid | - |
| [named-grid-areas-no-invalid](https://github.com/stylelint/stylelint/blob/14.8.2/lib/rules/named-grid-areas-no-invalid/README.md) | Yes | Mid | - |
|[no-invalid-position-at-import-rule](https://github.com/stylelint/stylelint/blob/14.8.2/lib/rules/no-invalid-position-at-import-rule/README.md) | Yes | Mid | - |
| [property-no-unknown](https://github.com/stylelint/stylelint/blob/14.8.2/lib/rules/property-no-unknown/README.md) | Yes | Mid | - |
| [selector-anb-no-unmatchable](https://github.com/stylelint/stylelint/blob/14.8.2/lib/rules/selector-anb-no-unmatchable/README.md) | Yes | Mid | - |
| [declaration-block-no-duplicate-properties](https://github.com/stylelint/stylelint/blob/14.8.2/lib/rules/declaration-block-no-duplicate-properties/README.md) | Yes | Mid | - |
| [no-descending-specificity](https://github.com/stylelint/stylelint/blob/14.8.2/lib/rules/no-descending-specificity/README.md) | Yes | Hard | - |
| [at-rule-no-unknown](https://github.com/stylelint/stylelint/blob/14.8.2/lib/rules/at-rule-no-unknown/README.md) | Yes | - | ~[parser completely covers this](https://biomejs.dev/playground/?files.main.css=QAB1AG4AawBuAG8AdwBuACAAewB9AAoACgA%3D)~ |
| [at-rule-no-unknown](https://github.com/stylelint/stylelint/blob/14.8.2/lib/rules/at-rule-no-unknown/README.md) | No | - | [parser completely covers this](https://biomejs.dev/playground/?files.main.css=QAB1AG4AawBuAG8AdwBuACAAewB9AAoACgA%3D) |
| [no-invalid-double-slash-comments](https://github.com/stylelint/stylelint/blob/14.8.2/lib/rules/no-invalid-double-slash-comments/README.md) | No | - | [parser covers this](https://biomejs.dev/playground/?files.main.css=YQAgAHsACgAgACAALwAvAGMAbwBsAG8AcgA6ACAAcABpAG4AawA7AAoAfQA%3D) |
| [annotation-no-unknown](https://github.com/stylelint/stylelint/blob/14.8.2/lib/rules/annotation-no-unknown/README.md)| ? | - | [parser reports error but a little imperfect](https://biomejs.dev/playground/?files.main.css=YQAgAHsACgAgACAAYwBvAGwAbwByADoAIABnAHIAZQBlAG4AIAAhAGkAbQBtAG0AbQBtAHAAbwByAHQAYQBuAHQAOwAKAH0A) |
| [color-no-invalid-hex](https://github.com/stylelint/stylelint/blob/14.8.2/lib/rules/color-no-invalid-hex/README.md) | ? | - | The parser covers some cases, but it's not perfect. Should we handle this in the parser or in the linter? |
| [function-calc-no-unspaced-operator](https://github.com/stylelint/stylelint/blob/14.8.2/lib/rules/function-calc-no-unspaced-operator/README.md) | ? | - | style rule? It would conflict with formatter? |
| [no-empty-source](https://github.com/stylelint/stylelint/blob/14.8.2/lib/rules/no-empty-source/README.md) | ? | - | syntax error in test case https://stylelint.io/user-guide/rules/no-empty-source/ |
| [string-no-newline](https://github.com/stylelint/stylelint/blob/14.8.2/lib/rules/string-no-newline/README.md) | ? | - | [style rule?](https://biomejs.dev/playground/?files.main.css=YQAgAHsACgAgACAAYwBvAG4AdABlAG4AdAA6ACAAIgBmAGkAcgBzAHQACgAgACAAIAAgAHMAZQBjAG8AbgBkACIAOwAKAH0A) |
| [media-query-no-invalid](https://github.com/stylelint/stylelint/blob/14.8.2/lib/rules/media-query-no-invalid/README.md) | ? | - | [parser will panic](https://biomejs.dev/playground/?files.main.css=QABtAGUAZABpAGEAIAAoAHcAaQBkAHQAaAAgAD0APQAgADEAMAAwAHAAeAApACAAewB9AA%3D%3D) |

**Comment by Th3S4mur41** [comment]
@togami2864 would you also consider adding [liberty/use-logical-spec](https://github.com/Jordan-Hall/stylelint-use-logical-spec) to the list although it is not a stylelint recommended rule?

Logical properties is a great way to improve UIs with various language directions but it is not easy to get devs to switch their habits from the old properties and something that should not be checked manually

**Reply by togami2864** [reply]
Hi @Th3S4mur41. Thank you for your suggesting.

Unfortunately, the core/maintainer team won't implement this right away. Sorting order is often quite controversial and developers have varied preferences, so we need to be very careful when adding it to the official rules.

If there’s a possibility, it's to implement it as a plugin. We already have a [workground](https://github.com/biomejs/biome/issues/2463), and it should be ready to use in the near future.
